### ISSUE TITLE
Button to invalidate all existing session

### ISSUE THREAD

**Issue by ekzyis** [stackernews/stacker.news#1017]
**Is your feature request related to a problem? Please describe.**

We don't have a button to invalidate all existing sessions.

**Describe the solution you'd like**

Button to invalidate all existing sessions. Could be in the settings.

**Describe alternatives you've considered**

n/a

**Additional context**

https://stacker.news/items/493900

**Comment by huumn** [comment]
We don't store sessions so I'm not sure how we'll manage to do this. It's all stateless.

**Comment by SatsAllDay** [comment]
Just thinking out loud here...

1. Add a button in settings that sets a `invalidateSessions` timestamp in the DB somewhere, maybe an `SessionInvalidationRequest` table.
2. Add middleware to nextjs that checks the DB for such an entry on requests when the user already has a session. If an entry is found, respond with a `Set-Cookie` header that clears the existing session cookie (the equivalent of a logout operation, I presume).
3. Somehow we'd have to only clear the session if it was created _before_ the latest `SessionInvalidationRequest` entry. Otherwise, you'd end up in an infinite loop where sessions are constantly revoked.

The first two steps seem reasonable, but the third I am not sure about.

**Comment by frstdrgn** [comment]
I didn't realize the sessions were stateless, I'll take a look at this and try to come up with some ideas.

> Add middleware to nextjs that checks the DB for such an entry on requests when the user already has a session. If an entry is found, respond with a Set-Cookie header that clears the existing session cookie (the equivalent of a logout operation, I presume).

Doesn't sound like this actually invalidates the cookie? Which in my mind is the security concern... 

This actually leaves with a question - do sessions expire at all?

**Comment by frstdrgn** [comment]
@ekzyis if sessions don't expire at all, I think that should be separate security issue fwiw

**Comment by ekzyis** [comment]
> Doesn't sound like this actually invalidates the cookie? Which in my mind is the security concern...

It would since any JWT that was created before user hit "invalidate all sessions" will no longer be accepted. So it's effectively invalidated, no?

> This actually leaves with a question - do sessions expire at all?

Default expiration is 30 days but they are regular refreshed while you're logged in and on the site.

**Comment by frstdrgn** [comment]
@ekzyis , clearing the session cookie via a Set-Cookie header doesn't invalidate the session, it simply removes it from the browser. See the screenshots, showing the "me" GraphQL query returning my user data while I'm logged in, the sign out request setting the cookie to be empty, and the same GraphQL query returning my user data again even AFTER the sign out request. You'll have to take my word for it that it's the same cookie (obviously don't want to send that in plain view), and take a look at the time stamps. You can also try this on your own by grabbing the cookie from your browser and placing it back in after logging out, or using Burp Suite repeater (which is free, but takes a few minutes to set up if you haven't used it before). 
![logged_in](https://github.com/stackernews/stacker.news/assets/166055143/ad567e1c-938c-4c52-910d-08df21908853)
![signout](https://github.com/stackernews/stacker.news/assets/166055143/4fbd46eb-39a2-4d50-99eb-26c7c3b34874)
![post-signout](https://github.com/stackernews/stacker.news/assets/166055143/a41115fb-7b93-43e5-bb9d-313e3c9a6229)

Other thing worth noting here is that the refresh request is being sent *every second*, which means *every second* a new valid session cookie is being generated and won't expire for 30 days. 

Personally I would recommend that be sent significantly less often, but these are all low or informational level risks.

**Comment by SatsAllDay** [comment]
Clearing the cookie via `Set-Cookie` alone doesn't invalidate the cookie. In my proposal, there would be additional requirements to track when sessions were requested to be invalidated, then some additional logic server-side to consult when the incoming jwt/cookie was generated, compare to the request to invalidate sessions, and conditionally allow/deny access accordingly. This approach does not work in today's codebase without modifications being made.

**Comment by ekzyis** [comment]
This sounds interesting:

> A common approach for invalidating tokens when a user changes their password is to sign the token with a hash of their password. Thus if the password changes, any previous tokens automatically fail to verify. You can extend this to logout by including a last-logout-time in the user's record and using a combination of the last-logout-time and password hash to sign the token. This requires a DB lookup each time you need to verify the token signature, but presumably you're looking up the user anyway.

-- stackoverflow.com, [_Invalidating JSON Web Tokens_](https://stackoverflow.com/questions/21978658/invalidating-json-web-tokens)

Instead of storing a date and having additional custom logic to check if a JWT was revoked, we could sign JWTs with unique keys per user. For example, we could generate some random data when a user is created and derive a new secret key from this random data + our secret in `NEXTAUTH_SECRET`. When they click on "invalidate all sessions", this secret would be updated.

If we would provide this derived [secret](https://next-auth.js.org/configuration/options#secret) here:

https://github.com/stackernews/stacker.news/blob/255ad29897361fec2338fcd7b7aef91d129bde97/pages/api/auth/%5B...nextauth%5D.js#L217-L234

This "more native solution" could "magically" "just work" then (assuming it's really that easy and there are no pitfalls).

**Comment by frstdrgn** [comment]
Yeah I think this is the simplest solution, and it sounds way easier than keeping a database of session tokens. This also gives you the option of resetting the key automatically every X days, so that a user can't refresh their session indefinitely. Can't see any meaningful risks with that idea as long as the user keys are generated via a CSPRNG and aren't exposed to the user.

**Comment by ekzyis** [comment]
> Would you need to override encode/decode if all you're doing is providing a different secret?

You are right, we don't afaict. No longer know what I was thinking in my previous comment lol

**Comment by frstdrgn** [comment]
Wanted to follow up here and see if there any questions I could help out with. I was thinking about this a bit more, and I think one of two options would work best here:

1. Issuing a unique key for each user as mentioned in @ekzyis's [comment here](https://github.com/stackernews/stacker.news/issues/1017#issuecomment-2046024492) - if an "invalidate all sessions" button is pressed, it resets that user's key and they have to reauth
2. Keeping track of a timestamp for each user that updates when they hit "invalidate all sessions". When validating JWTs, add a check that ensures the `iat` is greater than the timestamp

I don't think there necessarily has to be a huge rush on this, but I do think there should be a way for users to ensure sessions are revoked if they're concerned about a compromise, especially since you can refresh cookies indefinitely.
